ethcore/src/snapshot/service.rs

```diff
@@ -334,12 +338,144 @@ impl Service {
 
 	// replace one the client's database with our own.
 	fn replace_client_db(&self) -> Result<(), Error> {
-		let our_db = self.restoration_db();
+		let migrated_blocks = self.migrate_blocks()?;
+		trace!(target: "snapshot", "Migrated {} ancient blocks", migrated_blocks);
 
-		self.db_restore.restore_db(&*our_db.to_string_lossy())?;
+		let rest_db = self.restoration_db();
+		self.client.restore_db(&*rest_db.to_string_lossy())?;
 		Ok(())
 	}
 
+	// Migrate the blocks in the current DB into the new chain
+	fn migrate_blocks(&self) -> Result<usize, Error> {
+		// Count the number of migrated blocks
+		let mut count = 0;
+		let rest_db = self.restoration_db();
+
+		let cur_chain_info = self.client.chain_info();
+
+		let next_db = self.restoration_db_handler.open(&rest_db)?;
+		let next_chain = BlockChain::new(Default::default(), &[], next_db.clone());
+		let next_chain_info = next_chain.chain_info();
+
+		// Get the *inclusive* first and last blocks to iterate through
+		let start_block_number = match next_chain_info.ancient_block_number {
+			Some(bn) => bn + 1,
+			None => return Ok(count),
+		};
+		let mut end_block_number = match next_chain_info.first_block_number {
+			Some(bn) => bn - 1,
+			None => return Ok(count),
+		};
+
+		// Check that the intersection is non-null
+		if start_block_number >= end_block_number {
+			return Ok(count);
+		}
+
+		// Find the best block we can sync to
+		match (cur_chain_info.ancient_block_number, cur_chain_info.best_block_number) {
```

Comment:
Can you elaborate on the logic here a bit more? My understanding is that this is what the dbs look like:

```
current_chain:
[0, cur.ancient_block_number] ... [cur.first_block, cur.best_block]

new_chain:
[0, 0] ... [new.first_block, new.best_block]
```

So we need to import everything from the following ranges:
```
[0, min(cur.ancient_block_number, new.first_block)] if cur.ancient_block_number.is_some()
// or 
[0, min(cur.best_block, new.first_block)]
```

If that's the case we should try to simplify the logic here (and error handling especially - following this `match` statements makes it a bit hard to read).

I'd propose adding a closure that returns `-> Option<(from_hash, to_hash)>` and utilize `?` to early exit all `None`s, so the ccode would look like this:
```rust
/// Finds a range of blocks to import.
///
/// The old database looks like this:
/// [genesis, best_ancient_block] ... [first_block, best_block]
/// If we are fully synced neither `best_ancient_block` nor `first_block` is set, and we can assume that the whole range from [genesis, best_block] is imported.
/// The new database only contains the tip of the chain, so the useful set of blocks is defined as:
/// [0 ... min(new.first_block, best_ancient_block or best_block)]
fn find_range() -> Option<...> {
  let from = next_chain_info.ancient_block_number? + 1; // is this ever non 0?
  let to = next_chain_info.first_block_number? - 1;

  let available_to = cur_chain.ancient_block_number.unwrap_or(cur_chain_info.best_block_number);

  let to = cmp::min(to, available_to);

  if from >= to {
    return None;
  }

  let from_hash = self.client.block_hash(BlockId::Number(from - 1))?; //why -1?
  let to_hash = self.client.block_hash(BlockId::Number(to))?;

  Some((from_hash, to_hash))
}

```